File: ios/Sources/FlowbizOnsite/FlowbizPush.swift
import Foundation

/// Result of `Flowbiz.handlePush` (SPEC §10.2/§10.3): a decoded Flowbiz
/// push payload. `type` is free-form — new push kinds require no SDK
/// update; the SDK parses best-effort regardless of `version` (forward
/// compatibility).
public struct FlowbizPush: Sendable {

    /// Contract version (`v`); absent/malformed defaults to 1.
    public let version: Int
    /// Free-form push kind, e.g. `"cart_recovery"`. Always non-empty.
    public let type: String
    public let title: String?
    public let body: String?
    /// `deep_link` as a `URL`, or nil when absent or rejected by
    /// `URL(string:)` — the push itself is still returned then.
    public let deepLink: URL?
    /// `data` object of the decoded payload; empty when absent.
    public let data: [String: JSONValue]

    /// The raw `deep_link` string — kept internally so `recoveryPayload`
    /// works even when `URL(string:)` and the raw string disagree.
    let deepLinkString: String?

    /// Convenience for cart-recovery pushes (SPEC §10.2: the `mb_recovery`
    /// link rides in `deep_link`): the deep link run through the
    /// `Flowbiz.handleLink` decoder. Nil when there is no deep link or it
    /// carries no decodable `mb_recovery` value. Pure, like `handleLink`.
    public var recoveryPayload: RecoveryPayload? {
        RecoveryLinkParser.parse(deepLinkString)
    }
}

/// Pure parser behind `Flowbiz.handlePush` (SPEC §10.2): the value of the
/// `"flowbiz"` marker key — a JSON-encoded *string* (the cross-platform
/// contract; FCM data messages are flat maps) — decoded into a
/// `FlowbizPush`. APNs payloads are nested JSON, so an iOS sender *could*
/// put an object there; that is tolerated leniently (`parse(object:)`),
/// documented in `shared/push-samples/samples.json`.
///
/// Tolerant by design: unknown `v` values and unknown fields parse
/// best-effort (forward compatibility). Nil only for undecodable JSON, a
/// non-object root, or a missing/empty `type`. Never throws.
enum PushPayloadParser {

    static let markerKey = "flowbiz"

    static func parse(_ markerValue: String) -> FlowbizPush? {
        guard
            let root = try? JSONSerialization.jsonObject(with: Data(markerValue.utf8)),
            let payload = root as? [String: Any]
        else { return nil }
        return parse(object: payload)
    }

    static func parse(object payload: [String: Any]) -> FlowbizPush? {
        guard let type = payload["type"] as? String, !type.isEmpty else { return nil }
        let version: Int
        if let number = payload["v"] as? NSNumber, !JSONValue.isBoolean(number) {
            version = number.intValue
        } else {
            version = 1
        }
        let data = (payload["data"] as? [String: Any])
            .flatMap(JSONValue.objectFromFoundation) ?? [:]
        let deepLinkString = payload["deep_link"] as? String
        return FlowbizPush(
            version: version,
            type: type,
            title: payload["title"] as? String,
            body: payload["body"] as? String,
            deepLink: deepLinkString.flatMap(URL.init(string:)),
            data: data,
            deepLinkString: deepLinkString
        )
    }
}

File: android/sdk/src/main/kotlin/com/flowbiz/onsite/FlowbizPush.kt
package com.flowbiz.onsite

import android.net.Uri
import org.json.JSONObject

/**
 * Result of [Flowbiz.handlePush] (SPEC §10.2/§10.3): a decoded Flowbiz push
 * payload. [type] is free-form — new push kinds require no SDK update; the
 * SDK parses best-effort regardless of [version] (forward compatibility).
 *
 * Immutable. Not a data class on purpose: the raw deep-link string is an
 * internal field (the platform [Uri] is derived lazily so the type stays
 * constructible where `android.net.Uri` does not exist, e.g. JVM tests).
 */
class FlowbizPush internal constructor(
    /** Contract version (`v`); absent/malformed defaults to 1. */
    val version: Int,
    /** Free-form push kind, e.g. `"cart_recovery"`. Always non-empty. */
    val type: String,
    val title: String?,
    val body: String?,
    /** `data` object of the decoded payload; empty when absent. */
    val data: Map<String, Any?>,
    internal val deepLinkString: String?,
) {

    /**
     * `deep_link` as a [Uri], or null when absent. `Uri.parse` is lenient
     * (it validates nothing), so this is null only for an absent value; any
     * parse surprise degrades to null with the push still returned.
     */
    val deepLink: Uri?
        get() = try {
            deepLinkString?.let(Uri::parse)
        } catch (t: Throwable) {
            null
        }

    /**
     * Convenience for cart-recovery pushes (SPEC §10.2: the `_mb_cr_`
     * link rides in `deep_link`): the deep link run through the
     * [Flowbiz.handleLink] decoder. Null when there is no deep link or it
     * carries no decodable `_mb_cr_` value. Pure, like `handleLink`.
     */
    val recoveryPayload: RecoveryPayload?
        get() = RecoveryLinkParser.parse(deepLinkString)
}

/**
 * Pure parser behind [Flowbiz.handlePush] (SPEC §10.2): the value of the
 * `"flowbiz"` marker key — a JSON-encoded *string* (FCM data messages are
 * flat `Map<String, String>`) — decoded into a [FlowbizPush].
 *
 * Tolerant by design: unknown `v` values and unknown fields parse
 * best-effort (forward compatibility). Null only for undecodable JSON, a
 * non-object root, or a missing/empty `type`. Never throws.
 */
internal object PushPayloadParser {

    const val MARKER_KEY = "flowbiz"

    fun parse(markerValue: String?): FlowbizPush? {
        if (markerValue == null) return null
        return try {
            fromObject(JSONObject(markerValue))
        } catch (t: Throwable) {
            null
        }
    }

    private fun fromObject(payload: JSONObject): FlowbizPush? {
        val type = (payload.opt("type") as? String)?.takeIf { it.isNotEmpty() } ?: return null
        val version = (payload.opt("v") as? Number)?.toInt() ?: 1
        val data = payload.optJSONObject("data")?.let(JsonPlain::toPlainMap) ?: emptyMap()
        return FlowbizPush(
            version = version,
            type = type,
            title = payload.opt("title") as? String,
            body = payload.opt("body") as? String,
            data = data,
            deepLinkString = payload.opt("deep_link") as? String,
        )
    }
}
